wavgen.1
.TH wavgen 1 "October 1999" Linux "Linux Audio Manual"
.SH NAME
wavgen \- Test Wavefile Generator
.SH SYNOPSIS
.B wavgen [-v][-f <freq>][-l <level>][-c left|right|both][-p norm|rev]\
-t <secs> <wav-file>

.SH DESCRIPTION
\fBWavgen\fP is a command-line tool for producing audio test wavefiles, 
consisting of steady-state tones.

.SH OPTIONS
.TP
.B \-v
Print a version identifier and then exit.
.TP
.B \-f <freq>
Produce a file with a tone of frequency \fI<freq>\fP Hz.  If not specified, 
a frequency of 1000 Hz will be used by default.
.TP
.B \-l <level>
Produce a file with a tone at \fI<level>\fP dB below the maximum output level.
If not specified, a level of 0 dB will be used --i.e. maximum output level.
.TP
.B \-c left|right|both
Produce a file with a tone on the corresponding channel(s).  If not specified,
the tone will be recorded on both channels.
.TP
.B \-p norm|rev
If given as \fBnorm\fP, the tone will be recorded with both channels in phase.
If \fBrev\fP, then the channels will be 180 degrees out of phase.
.TP
.B \-t <secs>
Generate a file \fI<secs>\fP long.  

.SH AUTHOR
Fred Gleason <fredg@wava.com>
.SH SEE ALSO
.BR agen (1)


History
                    History file for the WAVLIB Project

-------------------------------------------------------------------------------
11/20/1999
Start of history file

Current released version is 2.2.2

Added the aproc(1) command, a command-line audio processor for wav files.
Currently, it only supports the 'normalize' function (-N), but is intended
as a catch-all for adding additional processing features as needed through
addition of additional command flags.

Added a man page for the aproc(1) command.

Fixed a bug in OpenWav() that was causing the file pointer to not be
pointing at the start of the wave data at function return with certain 
combinations of chunks. 

Fixed a typo in the ameter(1) man page.
-------------------------------------------------------------------------------

12/1/1999 -- v 2.3.1
Added a conditional definition of PI (definition is missing from glibc).
-------------------------------------------------------------------------------

5/18/2000 -- v 2.4.0

Cleaned up code in ameter(1) so that levels are now displayed in true 
logarithmic dB.

Added a '-l' option to ameter(1) to allow the user to set the meter reference
level

Added color to ameter(1), with a green and red segment.  Added a '-t' option
to allow the user to define where the border should be.

Updated the ameter(1) man page.

Fixed a bug in infowav(1) that was causing it to corrupt wav files when invoked
with the '-p' while omitting the audio-device name that customarily follows it.

Fixed a bug in infowav(1) that was causing extraneous output to be printed.

Added a '-t' option to recwav(1) to allow the user to specify a record length.

Updated the recwav(1) man page.

Added a TailTrim() API call to remove audio below a given set threshold level
from the end of an existing wav file.
-------------------------------------------------------------------------------

5/18/2000 -- v 2.4.1

Restored the RecordWavFile() API call, which somehow got dropped out of
v2.4.0.
-------------------------------------------------------------------------------

5/23/2000 -- v 2.4.2

Fixed a fencepost bug in TailTrim() that was causing false positive 
threshold detects on certain wav files.
-------------------------------------------------------------------------------

wavgen.c
/* wavgen.c
 *
 *  This puts a tone into a specified wav file.
 *
 *   (C) Copyright 2002-2005 Fred Gleason <fredg@salemradiolabs.com>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as
 *   published by the Free Software Foundation; either version 2 of
 *   the License, or (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public
 *   License along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <strings.h>
#include <unistd.h>
#include <sys/soundcard.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <math.h>
#include <time.h>
#include <string.h>

#include "wavlib.h"

#define BUFFER_SIZE 16384
#define SAMPLE_RATE 44100
#define SAMPLE_SIZE 16
#define USAGE "wavgen [-f freq][-l level][-c left|right|both][-p now|rev][-t secs][-v] <wavfile>\n"

int main(int argc,char *argv[])
{
  int i;
  int hFilename;
  unsigned char cBuffer[BUFFER_SIZE];
  float fFreq=1000;
  float fLevel=0;
  float fRatio;
  int dChan=0;     /* 0=Both, 1=Left, 2=Right */
  char sChan[10];
  int dPhase=0;    /* 0=Normal, 1=Reverse */
  char sPhase[10];
  int dTime=0;
  int dMatch=0;
  int dDebug=0;
  int dSampleRate=0;
  char sFilename[128];  /* Name of audio device */
  float fGain;
  float fAngle;
  unsigned long ldCount,ldTimeline=0,ldLimit=BUFFER_SIZE;
  short int dSample;
  time_t tmTimestamp,tmTime;
  struct wavHeader wavHeader;
  unsigned dBytes=0,dTotalBytes;

  if(argc<2) {
    printf(USAGE);
    exit(0);
  }

  if(argc==2) {
    if(strcmp(argv[1],"-v")==0) {
      printf("wavgen v");
      printf(VERNUM);
      printf("\n");
      exit(0);
    }
  }

  /* Get device name */
  if(sscanf(argv[argc-1],"%s",sFilename)!=1) {
    printf("wavgen: invalid file anem\n");
    exit(1);
  }

  /* Get options */
  for(i=1;i<argc-1;i++) {
    dMatch=0;
    if(strcmp(argv[i],"-v")==0) {     /* Version */
      printf("wavgen v");
      printf(VERNUM);
      printf("\n");
      exit(0);
    }
    if(strcmp(argv[i],"-d")==0) {     /* Debug Mode */
      dMatch=1;
      dDebug=1;
    }
    
    if(strcmp(argv[i],"-f")==0) {     /* Frequency */
      dMatch=1;
      if(sscanf(argv[++i],"%f",&fFreq)!=1) {
	printf("wavgen: invalid frequency\n");
	exit(1);
      }
    }
    if(strcmp(argv[i],"-l")==0) {    /* Level */
      dMatch=1;
      if(sscanf(argv[++i],"%f",&fLevel)!=1) {
	printf("wavgen: invalid level\n");
	exit(1);
      }
      fLevel=-fLevel;
    }
    if(strcmp(argv[i],"-t")==0) {    /* Time */
      dMatch=1;
      if(sscanf(argv[++i],"%d",&dTime)!=1) {
	printf("wavgen: invalid time interval\n");
	exit(1);
      }
    }
    if(strcmp(argv[i],"-c")==0) {    /* Channel */
      dMatch=1;
      if(sscanf(argv[++i],"%s",sChan)!=1) {
	printf("wavgen: invalid time interval\n");
	exit(1);
      }
      if(strcasecmp(sChan,"both")==0) {
	dChan=0;
      }
      else {
	if(strcasecmp(sChan,"left")==0) {
	  dChan=1;
	}
	else {
	  if(strcasecmp(sChan,"right")==0) {
	    dChan=2;
	  }
	  else {
	    printf("wavgen: invalid channel\n");
	    exit(1);
	  }
	}
      }  
    }
    if(strcmp(argv[i],"-p")==0) {    /* Phase */
      dMatch=1;
      if(sscanf(argv[++i],"%s",sPhase)!=1) {
	printf("wavgen: invalid phase setting\n");
	exit(1);
      }
      if(strcasecmp(sPhase,"norm")==0) {
	dPhase=0;
      }
      else {
	if(strcasecmp(sPhase,"rev")==0) {
	  dPhase=1;
	}
	else {
	  printf("wavgen: invalid phase setting\n");
	  exit(1);
	}
      }  
    }
    if(dMatch==0) {
      printf("wavgen: invalid option\n");
      exit(1);
    }
    if(dTime==0) {
      printf("wavgen: missing time argument\n");
      exit(1);
    }
  }

  /* Convert db to ratio */
  fRatio=pow(10,(fLevel/20));

  /* Set audio characteristics */
  wavHeader.wFormatTag=WAVE_FORMAT_PCM;
  wavHeader.wChannels=2;
  wavHeader.dwSamplesPerSec=SAMPLE_RATE;
  wavHeader.dwAvgBytesPerSec=SAMPLE_RATE*(SAMPLE_SIZE/8)*2;
  wavHeader.wBlockAlign=SAMPLE_SIZE/4;
  wavHeader.wBitsPerSample=SAMPLE_SIZE;
  
  /* Open the wav file */
  hFilename=CreateWav(sFilename,&wavHeader);
  if(hFilename<0) {
    printf("wavgen: can't open wav file\n");
    exit(1);
  }

  /* Display Settings (if requested) */
  if(dDebug==1) {
    printf("--Audio Generator Settings--\n");
    printf("Frequency: %5.0f Hz\n",fFreq);
    printf("Level: %3.1f dB\n",fLevel);
    printf("Channel(s): ");
    switch(dChan) {
    case 0:
      printf("BOTH\n");
      break;
    case 1:
      printf("LEFT\n");
      break;
    case 2:
      printf("RIGHT\n");
      break;
    }
    printf("Phasing: ");
    switch(dPhase) {
    case 0:
      printf("NORMAL\n");
      break;
    case 1:
      printf("REVERSE\n");
      break;
    }
    printf("Effective Sample Rate: %d samples/sec\n",dSampleRate);
  }

  /* Setup time data */
  time(&tmTimestamp);
  tmTime=tmTimestamp;
  if(dTime>0) {
    tmTimestamp+=dTime;
  }
  else {
    tmTimestamp--;
  }

  /* Output audio */
  dTotalBytes=wavHeader.dwAvgBytesPerSec*dTime;

  switch(SAMPLE_SIZE) {
  case 8:
    fGain=127*fRatio;
    ldTimeline=0;
    ldLimit=BUFFER_SIZE/2;
    while(dBytes<dTotalBytes) { 
      i=0;
      for(ldCount=ldTimeline;ldCount<ldLimit;ldCount++) {
	fAngle=2*PI*ldCount*fFreq/SAMPLE_RATE;
	if(dChan==0 || dChan==1) {
	  cBuffer[i++]=fGain*sin(fAngle)+128;
	}
	else {
	  cBuffer[i++]=128;
	}
	if(dChan==0 || dChan==2) {
	  if(dPhase==0) {
	    cBuffer[i++]=fGain*sin(fAngle)+128;
	  }
	  else {
	    cBuffer[i++]=-fGain*sin(fAngle)+128;
	  }
	}
	else {
	  cBuffer[i++]=128;
	}
      }
      write(hFilename,cBuffer,BUFFER_SIZE);
      dBytes+=BUFFER_SIZE;
      ldLimit+=(BUFFER_SIZE/2);
      ldTimeline+=(BUFFER_SIZE/2);
      time(&tmTime);
    }  
    break;
  case 16:
    fGain=32767*fRatio;
    ldTimeline=0;
    ldLimit=BUFFER_SIZE/4;
    while(dBytes<dTotalBytes) { 
      i=0;
      for(ldCount=ldTimeline;ldCount<ldLimit;ldCount++) {
	fAngle=2*PI*ldCount*fFreq/SAMPLE_RATE;
	dSample=fGain*sin(fAngle);
	if(dChan==0 || dChan==1) {
	  cBuffer[i++]=0xFF&dSample;
	  cBuffer[i++]=(0xFF&(dSample>>8));  
	}
	else {
	  cBuffer[i++]=0;
	  cBuffer[i++]=0;
	}
	if(dChan==0 || dChan==2) {
	  if(dPhase==1) {
	    dSample=-dSample;
	  }
	  cBuffer[i++]=0xFF&dSample;
	  cBuffer[i++]=(0xFF&(dSample>>8));
	} 
	else {
	  cBuffer[i++]=0;
	  cBuffer[i++]=0;
	}
      }
      write(hFilename,cBuffer,BUFFER_SIZE);
      dBytes+=BUFFER_SIZE;
      ldLimit+=(BUFFER_SIZE/4);
      ldTimeline+=(BUFFER_SIZE/4);
      time(&tmTime);
    }  
  break;
  }

  /* close files and finish */
  FixWav(hFilename,dBytes/wavHeader.wBlockAlign,dBytes);
  close(hFilename);


  exit(0);
}
